src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCSVFileRecord.java

```diff
@@ -516,4 +523,58 @@ public boolean next() throws SQLServerException {
         }
         return (null != currentLine);
     }
+
+    /**
+     * Returns whether the rules to escape delimiters are used.
+     *
+     * @return true if the rules are used, false otherwise.
+     */
+    public boolean isEscapeColumnDelimitersCSV() {
+        return escapeDelimiters;
+    }
+
+    /**
+     * When set to true, the following rules will be used to parse CSV files: Each field may or may not be enclosed in
+     * double quotes. If fields are not enclosed with double quotes, then double quotes may not appear inside the
+     * fields. Fields containing double quotes, and delimiters should be enclosed in double quotes. If double-quotes are
+     * used to enclose fields, then a double-quote appearing inside a field must be escaped by preceding it with another
+     * double quote.
+     *
+     * @param escapeDelimiters
+     *        true if the rules above to be used.
+     */
+    public void setEscapeColumnDelimitersCSV(boolean escapeDelimiters) {
+        this.escapeDelimiters = escapeDelimiters;
+    }
+
+    private static String[] escapeQuotesRFC4180(String[] tokens) {
+        if (null == tokens) {
+            return tokens;
+        }
+        for (int i = 0; i < tokens.length; i++) {
+            boolean escaped = false;
+            int j = 0;
+            StringBuilder sb = new StringBuilder();
+            if (tokens[i].contains("\"")) {
+                tokens[i] = tokens[i].trim();
+            }
+            while (j < tokens[i].length()) {
+                if ('"' == tokens[i].charAt(j)) {
+                    if (!escaped) {
+                        escaped = true;
+                    } else {
+                        if ((j < tokens[i].length() - 1) && '"' == tokens[i].charAt(j + 1)) {
```

Comment:
I think we should add an else statement to this if, to handle the case where the CSV data is corrupted and has a double quote that has not been escaped.

Reply:
Let me know if you'd like to leave the code as it is here as well, we can make the call as to how we're going to handle invalid cases. I think being lenient and letting un-closed double quotes in (the code is removing the un-closed double quote) works for me.

Reply:
That's a good catch. I will modify to throw an exception when there are odd number of quotes.

Reply:
not a bit deal...but typically these kind of errors give an error like "Unclosed quotation mark after the character string 'blah'"